src/lib/mina.ts

```diff
@@ -1,214 +1,306 @@
+    const pk = k.toPublicKey();
+    addAccount(pk, largeValue);
+    testAccounts.push({ privateKey: k, publicKey: pk });
   }
-}
-
-export type Amount = UInt32;
-export type Nonce = UInt32;
 
-export class MerkleProof extends CircuitValue {}
+  const getAccount = (pk: PublicKey): Promise<Account> => {
+    const r = ledger.getAccount(pk);
+    if (r == null) {
+      throw new Error(
+        `getAccount: Could not find account for ${JSON.stringify(pk.toJSON())}`
+      );
+    } else {
+      const a = {
+        balance: new UInt64(r.balance.value),
+        nonce: new UInt32(r.nonce.value),
+        snapp: r.snapp,
+      };
+      return new Promise((r) => r(a));
+    }
+  };
 
-export class MerkleCollection<T extends CircuitValue> {
-  get(_p: MerkleProof): T {
-    throw 'unimplemented';
+  function epochData(d: EpochDataPredicate) {
+    return {
+      ledger: {
+        hash: d.ledger.hash_,
+        totalCurrency: d.ledger.totalCurrency,
+      },
+      seed: d.seed_,
+      startCheckpoint: d.startCheckpoint_,
+      lockCheckpoint: d.lockCheckpoint_,
+      epochLength: d.epochLength,
+    };
   }
 
-  constructor(xs: () => T[]) {
-    console.log(xs);
-  }
-}
+  const body = (b: Body): SnarkyBody => {
+    return {
+      publicKey: b.publicKey,
+      update: b.update,
+      tokenId: b.tokenId,
+      delta: b.delta,
+      // TODO: events
+      events: [],
+      sequenceEvents: [],
+      // TODO: calldata
+      callData: Field.zero,
+      // TODO
+      depth: 0,
+    };
+  };
 
-export enum Permission {
-  NoAuthRequired,
-  Signature,
-  Proof,
-}
+  const transaction = (
+    sender: PrivateKey,
+    f: () => void | Promise<void>
+  ): Transaction => {
+    if (currentTransaction !== undefined) {
+      throw new Error(
+        'Cannot start new transaction within another transaction'
+      );
+    }
 
-export type Permissions = {
-  receive: Permission;
-  send: Permission;
-  setDelegate: Permission;
-  modifyState: Permission;
-};
+    currentTransaction = {
+      sender,
+      parties: [],
+      nextPartyIndex: 0,
+      protocolState: ProtocolStatePredicate.ignoreAll(),
+    };
+
+    const result = Circuit.runAndCheck(() => {
+      const res = f();
+      if (res instanceof Promise) {
+        return res.then(() => {
+          return () => {};
+        });
+      } else {
+        const r: Promise<() => void> = new Promise((k) => k(() => {}));
+        return r;
+      }
+    }).catch((err) => {
+      currentTransaction = undefined;
+      // TODO would be nice if the error would be a bit more descriptive about what failed
+      throw err;
+    });
+
+    const senderPubkey = sender.toPublicKey();
+    const txn = result
+      .then(() => getAccount(senderPubkey))
+      .then((senderAccount) => {
+        if (currentTransaction === undefined) {
+          throw new Error('Transaction is undefined');
+        }
+
+        const otherParties: Array<SnarkyParty> = currentTransaction.parties.map(
```

Comment:
I think we generally shouldn't use type annotations in assignments, except where necessary. TS infers these types, just like Ocaml. The benefits of type inference are that
* it is more flexible than prescribing a type
* less types have to be imported between files, leading to simpler code